firedrake/mg/mesh.py

```diff
@@ -80,28 +84,53 @@ def __getitem__(self, idx):
 
 def MeshHierarchy(mesh, refinement_levels,
                   refinements_per_level=1,
+                  degree=1,
                   reorder=None,
                   distribution_parameters=None, callbacks=None,
                   mesh_builder=firedrake.Mesh):
     """Build a hierarchy of meshes by uniformly refining a coarse mesh.
 
-    :arg mesh: the coarse :func:`~.Mesh` to refine
-    :arg refinement_levels: the number of levels of refinement
-    :arg refinements_per_level: the number of refinements for each
-        level in the hierarchy.
-    :arg distribution_parameters: options controlling mesh
-        distribution, see :func:`~.Mesh` for details.  If ``None``,
-        use the same distribution parameters as were used to
-        distribute the coarse mesh, otherwise, these options override
-        the default.
-    :arg reorder: optional flag indicating whether to reorder the
-         refined meshes.
-    :arg callbacks: A 2-tuple of callbacks to call before and
+    Parameters
+    ----------
+    mesh : :py:func:`~.Mesh`
+        the coarse mesh to refine
+    refinement_levels : int
+        the number of levels of refinement
+    refinements_per_level : int
+        the number of refinements for each level in the hierarchy.
+    degree : int
+        the degree of the finite element space used for
+        isoparametric representation of the geometry.
+        Only relevant if mesh is a Netgen mesh.
+    distribution_parameters : dict
+        options controlling mesh distribution, see :py:func:`~.Mesh`
+        for details.  If ``None``, use the same distribution
+        parameters as were used to distribute the coarse mesh,
+        otherwise, these options override the default.
+    reorder : bool
+        optional flag indicating whether to reorder the
+        refined meshes.
+    callbacks : tuple
+        A 2-tuple of callbacks to call before and
         after refinement of the DM.  The before callback receives
         the DM to be refined (and the current level), the after
         callback receives the refined DM (and the current level).
-    :arg mesh_builder: Function to turn a DM into a ``Mesh``. Used by pyadjoint.
+    mesh_builder
+        Function to turn a DM into a ``Mesh``. Used by pyadjoint.
+    Returns
+    -------
+    A :py:class:`HierarchyBase` object representing the
+    mesh hierarchy.
     """
+    if netgen and hasattr(mesh, 'netgen_mesh'):
```

Comment:
The issue with this is that behaviour silently changes depending on what is installed. I think we should get rid of the silent dispatch to NetgenHierarchy here and instead have a parameter to the MeshHeirarchy constructor to select whether to use a Firedrake intrinsic hierarchy or a netgen one. Choosing netgen when it's not installed would then trigger the error.

Reply:
It could then also be fatal to specify the degree if you're not using netgen.